File: tesisanalytics/src/main/scala/CommunicationLayerPython/Messages.scala
package CommunicationLayerPython

import java.util.UUID

import upickle.default.{ReadWriter => RW, macroRW}


trait Message

sealed trait Request extends Message

object Request{
  implicit val rw: RW[Request] = macroRW

}
sealed trait Response extends Message


case class ActionsForState(actions: Int)

case class Regard(value: Double)


//Initial Welcome
case class Welcome(message: String) extends Response
case class Farewell(message: String) extends Response


//********************* REQUEST ********************************


/**
  * Asks for a new experiment
  * @param experiment: Experiment TAG
  */
case class NewExperiment(experiment: String) extends Request
/**
  * Alter the experiment with action
  * @param experimentID
  * @param action: index of the action
  */
case class Action(experimentID: UUID, action: Int) extends Request

/**
  * Request to end the experimentID
  * @param experimentID
  */
case class EndExperiment(experimentID: UUID) extends Request




//********************* RESPONSE ********************************

case class InvalidRequest(error: String) extends Response

/**
  *
  * @param experimentID
  * @param actionsForState, the initial actions for the initia  state
  */
case class ExperimentResp(experimentID: UUID,
                      actionsForState: ActionsForState) extends Response

/**
  * the enviroment is updated, only these new actions can be done
  * @param actionsForNewState
  * @param regard: the regard for the previous action
  */
case class NewState(actionsForNewState: ActionsForState,
                    regard: Regard) extends Response

object NewExperiment{
  implicit val rw: RW[NewExperiment] = macroRW
}
object Action{
  implicit val rw: RW[Action] = macroRW
}
object Regard{
  implicit val rw: RW[Regard] = macroRW
}
object ActionsForState{
  implicit val rw: RW[ActionsForState] = macroRW
}
object ExperimentResp{
  implicit val rw: RW[ExperimentResp] = macroRW
}
object NewState{
  implicit val rw: RW[NewState] = macroRW
}
object InvalidRequest{
  implicit val rw: RW[InvalidRequest] = macroRW
}
object Welcome{
  implicit val rw: RW[Welcome] = macroRW
}
object Farewell{
  implicit val rw: RW[Farewell] = macroRW
}
object EndExperiment{
  implicit val rw: RW[EndExperiment] = macroRW
}

File: ejerl/src/main/python/messages.py
import json


class ActionsForState(object):
    def __init__(self, actions):
        self.actions = actions

    @staticmethod
    def readfromstr(dictParam):
        return ActionsForState(dictParam["actions"])

    def asJson(self):
        return {"actions": self.actions}


class Regard(object):
    def __init__(self, value):
        self.value = value

    @staticmethod
    def readfromstr(dictParam):
        return Regard(dictParam["value"])

    def asJson(self):
        return {"value": self.value}


class Message(object):
    def __init__(self):
        pass

    @staticmethod
    def parseMessage(val):
        j = json.loads(val)
        typeClass = j["$type"]

        if typeClass == Welcome.typeClass:
            return Welcome(j["message"])

        if typeClass == InvalidRequest.typeClass:
            return InvalidRequest(j["error"])

        if typeClass == ExperimentResp.typeClass:
            return ExperimentResp(j["experimentID"], j["stateDim"])

        if typeClass == Farewell.typeClass:
            return Farewell(j["message"])

        if typeClass == NewExperiment.typeClass:
            return NewExperiment(j["experiment"])

        if typeClass == Action.typeClass:
            return Action(j["experimentID"], j["action"])

        if typeClass == NewState.typeClass:
            return NewState(j["state"],j["actionsForNewState"], j["regard"])

        raise NotImplementedError("can not parse " + str(j))


class Request(Message):
    def __init__(self):
        super().__init__()


class Response(Message):
    def __init__(self):
        super().__init__()


class Welcome(Message):
    typeClass = "CommunicationLayerPython.Welcome"

    def __init__(self, message):
        self.message = message

    def __str__(self):
        var = json.dumps({"$type": Welcome.typeClass,
                          "message": self.message})
        return str(var)


class Farewell(Message):
    typeClass = "CommunicationLayerPython.Farewell"

    def __init__(self, message):
        self.message = message

    def __str__(self):
        var = json.dumps({"$type": Farewell.typeClass,
                          "message": self.message})
        return str(var)


# Requests


class NewExperiment(Request):
    typeClass = "CommunicationLayerPython.NewExperiment"

    def __init__(self, experiment):
        self.experiment = experiment

    def __str__(self):
        var = json.dumps({"$type": NewExperiment.typeClass,
                          "experiment": self.experiment})
        return str(var)




class Action(Request):
    typeClass = "CommunicationLayerPython.Action"

    def __init__(self, experimentID, action):
        self.experimentID = experimentID
        self.action = action

    def __str__(self):
        var = json.dumps({"$type": Action.typeClass,
                          "experimentID": self.experimentID,
                          "action": self.action})
        return str(var)


class EndExperiment(Request):
    typeClass = "CommunicationLayerPython.EndExperiment"

    def __init__(self, experimentID):
        self.experimentID = experimentID

    def __str__(self):
        var = json.dumps({"$type": EndExperiment.typeClass,
                          "experimentID": self.experimentID})
        return str(var)


class ResetExperiment(Request):
    typeClass = "CommunicationLayerPython.ResetExperiment"

    def __init__(self, experimentID):
        self.experimentID = experimentID

    def __str__(self):
        var = json.dumps({"$type": ResetExperiment.typeClass,
                          "experimentID": self.experimentID})
        return str(var)


# Response


class InvalidRequest(Response):
    typeClass = "CommunicationLayerPython.InvalidRequest"

    def __init__(self, error):
        self.error = error

    def __str__(self):
        var = json.dumps({"$type": InvalidRequest.typeClass,
                          "error": self.error})
        return str(var)


class ExperimentResp(Response):
    typeClass = "CommunicationLayerPython.ExperimentResp"

    def __init__(self, experimentID, stateDim):
        self.experimentID = experimentID
        self.stateDim = stateDim

    def __str__(self):
        var = json.dumps({"$type": ExperimentResp.typeClass,
                          "experimentID": self.experimentID,
                          "stateDim": self.stateDim})
        return str(var)


class NewState(Response):
    typeClass = "CommunicationLayerPython.NewState"

    def __init__(self,state, actionsForNewState, regard):
        self.state = state
        self.actionsForNewState = ActionsForState.readfromstr(actionsForNewState)
        self.regard = Regard.readfromstr(regard)

    def __str__(self):
        var = json.dumps({"$type": NewState.typeClass,
                          "state": self.state,
                          "actionsForNewState": self.actionsForNewState.asJson(),
                          "regard": self.regard.asJson()})
        return str(var)
